### PR TITLE
Added a cookie for authentication process

### PR DESCRIPTION
**Pull request by tharindulak**

### CODE REVIEW

File: components/api/constants/constants.bal [hunk 1]
```diff
@@ -36,3 +36,5 @@ public const string IDP_DEFAULT_PASSWORD = "admin";
 public const string IDP_DEFAULT_USERNAME = "admin";
 
 public const string AUTHORIZATION_HEADER = "Authorization";
+public const string COOKIE_HEADER = "Cookie";
+public const string COOKIE_KEY = "chpat";
```
Comment (nadundesilva) [inline components/api/constants/constants.bal]: Lets make update this to TOKEN_COOKIE_KEY to be more meaningful

File: components/api/filter/validator-filter.bal [hunk 2]
```diff
@@ -35,14 +35,22 @@ public type validateRequestFilter object {
             request.removeHeader(constants:AUTHENTICATED_USER);
         }
         string token = "";
-        if(request.hasHeader(constants:AUTHORIZATION_HEADER)) {
+        if(request.hasHeader(constants:AUTHORIZATION_HEADER) && request.hasHeader(constants:COOKIE_HEADER)) {
             string tokenHeaderValue = request.getHeader(constants:AUTHORIZATION_HEADER);
             string[] splittedToken = tokenHeaderValue.split(" ");
             if splittedToken.length() != 2 || !(splittedToken[0] == "Bearer" || splittedToken[0] == "bearer") {
-                log:printError("Not not receive the token in proper format");
+                log:printError("Did not the token in proper format");
```
Comment (nadundesilva) [inline components/api/filter/validator-filter.bal]: Please recheck this log. There seem to be a mistake here

File: components/api/filter/validator-filter.bal [hunk 3]
```diff
@@ -35,14 +35,22 @@ public type validateRequestFilter object {
...
                 return true;
             }
-            token = splittedToken[1];
+            string tokenLast = splittedToken[1];
+            string cookieHeader = request.getHeader(constants:COOKIE_HEADER);
+            string|error tokenFirst = getCookie(cookieHeader);
+            if tokenFirst is error {
+                log:printError("Cookied value could not be resolved. Passing to the next filter", err = tokenFirst);
```
Comment (nadundesilva) [inline components/api/filter/validator-filter.bal]: Typo in log

File: components/api/filter/validator-filter.bal [hunk 4]
```diff
@@ -35,14 +35,22 @@ public type validateRequestFilter object {
...
             if splittedToken.length() != 2 || !(splittedToken[0] == "Bearer" || splittedToken[0] == "bearer") {
-                log:printError("Not not receive the token in proper format");
+                log:printError("Did not the token in proper format");
                 return true;
             }
-            token = splittedToken[1];
+            string tokenLast = splittedToken[1];
+            string cookieHeader = request.getHeader(constants:COOKIE_HEADER);
+            string|error tokenFirst = getCookie(cookieHeader);
+            if tokenFirst is error {
```
Comment (nadundesilva) [inline components/api/filter/validator-filter.bal]: Please add brackets around if condition

File: components/api/filter/validator-filter.bal [hunk 5]
```diff
@@ -35,14 +35,22 @@ public type validateRequestFilter object {
             request.removeHeader(constants:AUTHENTICATED_USER);
         }
         string token = "";
-        if(request.hasHeader(constants:AUTHORIZATION_HEADER)) {
+        if(request.hasHeader(constants:AUTHORIZATION_HEADER) && request.hasHeader(constants:COOKIE_HEADER)) {
             string tokenHeaderValue = request.getHeader(constants:AUTHORIZATION_HEADER);
             string[] splittedToken = tokenHeaderValue.split(" ");
             if splittedToken.length() != 2 || !(splittedToken[0] == "Bearer" || splittedToken[0] == "bearer") {
```
Comment (sinthuja) [inline components/api/filter/validator-filter.bal]: Don't we have a way to check equalsIngnoreCase?

File: components/api/filter/validator-filter.bal [hunk 6]
```diff
@@ -80,3 +88,29 @@ public type validateRequestFilter object {
         return true;
     }
 };
+
+function getCookie(string cookiesString) returns string|error {
+    string cookieValue = "";
+    string[] cookies = cookiesString.split("[ ]{1,}");
```
Comment (nadundesilva) [inline components/api/filter/validator-filter.bal]: Isnt it easier to change this regex to "\\s*;\\s*"

File: components/api/filter/validator-filter.bal [hunk 7]
```diff
@@ -35,14 +35,22 @@ public type validateRequestFilter object {
...
+        if(request.hasHeader(constants:AUTHORIZATION_HEADER) && request.hasHeader(constants:COOKIE_HEADER)) {
             string tokenHeaderValue = request.getHeader(constants:AUTHORIZATION_HEADER);
             string[] splittedToken = tokenHeaderValue.split(" ");
             if splittedToken.length() != 2 || !(splittedToken[0] == "Bearer" || splittedToken[0] == "bearer") {
-                log:printError("Not not receive the token in proper format");
+                log:printError("Did not the token in proper format");
                 return true;
             }
-            token = splittedToken[1];
+            string tokenLast = splittedToken[1];
```
Comment (sinthuja) [inline components/api/filter/validator-filter.bal]: rename this to lastTokenElement.

File: components/api/filter/validator-filter.bal [hunk 8]
```diff
@@ -35,14 +35,22 @@ public type validateRequestFilter object {
...
+            string cookieHeader = request.getHeader(constants:COOKIE_HEADER);
+            string|error tokenFirst = getCookie(cookieHeader);
```
Comment (sinthuja) [inline components/api/filter/validator-filter.bal]: Rename this to firstTokenElement

File: components/api/filter/validator-filter.bal [hunk 9]
```diff
@@ -35,14 +35,22 @@ public type validateRequestFilter object {
...
+            if tokenFirst is error {
+                log:printError("Cookied value could not be resolved. Passing to the next filter", err = tokenFirst);
+                return true;
+            } else{
```
Comment (sinthuja) [inline components/api/filter/validator-filter.bal]: Formatting error.

File: components/api/filter/validator-filter.bal [hunk 10]
```diff
@@ -35,14 +35,22 @@ public type validateRequestFilter object {
...
+            string tokenLast = splittedToken[1];
+            string cookieHeader = request.getHeader(constants:COOKIE_HEADER);
+            string|error tokenFirst = getCookie(cookieHeader);
+            if tokenFirst is error {
+                log:printError("Cookie value could not be resolved. Passing to the next filter", err = tokenFirst);
+                return true;
+            } else{
+                token = io:sprintf("%s%s", tokenFirst, tokenLast);
+            }
             if "" == token {
```
Comment (sinthuja) [inline components/api/filter/validator-filter.bal]: Token is assigned in line #52, and hence check whether it's empty in that else section it self, rather having another if block.

**Pull request merged**